### PR TITLE
add close_range wrapper

### PR DESCRIPTION
**Pull request by vanym**
We need to wrap this new Linux call to avoid closing of systemd sockets. New version of ssh already uses it, without this fix hack didn't work.

### CODE REVIEW

**Comment by ryancdotorg** [comment]
Thanks for this. I'm looking at the logic for this, and it's a bit hard to follow. Let me see if I can restructure it.

I'm going to use `ex_first` and `ex_last` for first and last excluded fds in this pseudocode:

```
# ranges do not intersect
if (
  ex_first > ex_last or
  first > last or
  last < ex_first or
  first > ex_last
):
    return _real_close_range(first, last, flags)

# we know the ranges intersect
# is there a non-intersecting sub-range before the excluded range?
if first < ex_first:
    if ret := _real_close_range(first, ex_first - 1, flags):
        return ret

# is there a non-intersecting sub-range after the excluded range?        
if last > ex_last:
    if ret := _real_close_range(ex_last + 1, last, flags):
        return ret

# XXX only needed for debug
# was the range specified entirely within the excluded range?
if first >= ex_first and last <= ex_last:
    pass

return 0
```

Does this look right to you?

**Comment by ryancdotorg** [comment]
I'm also pondering whether a fallback function should be added for cases where `close_range` isn't available. I could update the DLWRAP macro to allow a fallback function to be provided, which at the very least could call `abort()`.

**Comment by ryancdotorg** [comment]
`close_range` seems possible, but really annoying to simulate - I will just add code to DLWRAP which prints an error message calls `abort()` if the target function isn't found, don't worry about that.

**Comment by vanym** [comment]
> I'm also pondering whether a fallback function should be added for cases where close_range isn't available. I could update the DLWRAP macro to allow a fallback function to be provided, which at the very least could call abort().

A scenario in which someone manages to call our `close_range` without existence of original seems unlikely, but possible

**Comment by ryancdotorg** [comment]
I think it would only happen if the software calling it was, itself, using `dlsym` to find the function, or perhaps if the binary were compiled on a newer system and run on an older one.

Ideally, we should try to continue in such a case, but otherwise, printing an error and calling abort is way better and easy to implement.

**Comment by ryancdotorg** [comment]
Thank you for the revised PR, that's much better!

I'd still like a range check on `sd_min_fd` and `sd_last_fd`.

I can merge your PR as-is and add the check myself in a separate commit, or you can add it - which would you prefer?

I also considered whether the "range entirely inside" bit should be in an `#ifdef NDEBUG` block, but with optimizations enabled the compiler should automatically exclude it regardless. I'll check in Compiler Explorer (godbolt.org) to be sure.

I'm also really glad to see someone else has found this hack useful!

**Comment by vanym** [comment]
> I'd still like a range check on `sd_min_fd` and `sd_last_fd`.

You mean you want variables names inside `wrap_close_range` to called so?
I can do `usd_min_fd` and `usd_last_fd` like it was before.
We have to reassign variables to `unsigned int` to avoid warnings or add warnings ignore rule.

**Comment by vanym** [comment]
> I'd still like a range check on `sd_min_fd` and `sd_last_fd`.

Oh, yes, in `ranges do not intersect` `if` we need to check them signed because `sd_last_fd` is `-1` by default.

**Comment by ryancdotorg** [comment]
Merged, thank you.

**Comment by vanym** [comment]
You changed `flags` arg from `int` to `unsigned int`. It was `int` on purpose. Function in `unistd.h` uses `int`. There is miss match between syscall and lib function. It won't compile.
```
gcc -O2 --std=gnu11 -Wall -Wextra -Wno-unused-function -shared -fPIC -s -DNDEBUG sdsock.c addr.c -lsystemd -ldl -o libsdsock.so
In file included from sdsock.c:24:
sdsock.c:39:8: error: conflicting types for ‘close_range’; have ‘int(unsigned int,  unsigned int,  unsigned int)’
   39 | DLWRAP(close_range, int, (unsigned int a, unsigned int b, unsigned int c), (a, b, c));
      |        ^~~~~~~~~~~
dlwrap.h:38:6: note: in definition of macro ‘DLWRAP’
   38 | TYPE NAME ARGS { return _impl_##NAME ARGN; }
      |      ^~~~
In file included from sdsock.c:6:
/usr/include/unistd.h:1208:12: note: previous declaration of ‘close_range’ with type ‘int(unsigned int,  unsigned int,  int)’
 1208 | extern int close_range (unsigned int __fd, unsigned int __max_fd,
      |            ^~~~~~~~~~~
make: *** [Makefile:25: libsdsock.so] Error 1
```

**Comment by ryancdotorg** [comment]
Ugh, sorry, that's annoying. I built it on an older system that doesn't have close_range so the error wasn't triggered. I'll fix that.

**Comment by ryancdotorg** [comment]
@vanym Try now?

**Comment by vanym** [comment]
> Try now?

It compiles, but you forget to change `%u` to `%d`.

Also I think discarding `ret` variable not worth it, we are doing hack that replaces original functions, who knows, maybe someone will pass non `-1` as error.

**Comment by vanym** [comment]
About `closefrom`. Wrapper looks wrong, it should call `_real_closefrom(sd_last_fd + 1)` after all. Also there is should be debug output like in other wrappers. I'll do that fix

It was my next step to build libsdsock in alpine docker. There build problem in alpine. It uses `elogind` instead of `systemd`, I'll research this too.

**Comment by ryancdotorg** [comment]
You're right about `closefrom`, please check the fix I pushed?

**Comment by vanym** [comment]
> I'll do that fix

I did in my [close-range branch](https://github.com/ryancdotorg/libsdsock/compare/main...vanym:libsdsock:close-range) 956f62b. Looks like Github doesn't notify about that here since it already merged.

Meanwhile you also try to fix it in e4038eb "more warnings" commit and here 8605e95. It also incorrect, if you pass number 3 to it first check will fail, second also fail and it will do nothing

**Comment by vanym** [comment]
Oh, you already did change it in third b2992ea commit. Looks like it'll work correctly.

Why you use "calling ... instead of ..." in `wrap_close_range` but did it differently in `wrap_closefrom`. Also there is no debug output about closing fds before sd_min_fd.

I think `errno = 0`  should be conditionally, because it's unlikely (btw I did `unlucky` misstypo in code comment) to happen and why to change it without need inside `closefrom` that do not change it by itself

**Comment by ryancdotorg** [comment]
I cherry-picked in your commit to incorporate those changes. I also modified the code to save/restore errno instead of just clearing it, just in case the caller is making assumptions.

**Comment by ryancdotorg** [comment]
Can you please make an issue for whatever needs to be done for alpine? I'm a little confused since I didn't think it used systemd?

**Comment by vanym** [comment]
> I cherry-picked in your commit to incorporate those changes. I also modified the code to save/restore errno instead of just clearing it, just in case the caller is making assumptions.

Okay, thanks for compromise
